language-plutus-core/stdlib/Language/PlutusCore/StdLib/Type.hs

```diff
@@ -1,43 +1,588 @@
+
+    \(a :: *) -> ifix
+        (\(rec :: ((* -> *) -> *) -> *) -> \(spine :: (* -> *) -> *) ->
+            spine (\(a :: *) -> all (r :: *). r -> (a -> rec (\(dat :: * -> *) -> dat a) -> r) -> r))
+        (\(dat :: * -> *) -> dat a)
+
+But we can't just normalize everything, because the user might write a non-normalized type and it's
+desirable to preserve redexes in the type.
+
+Then the question is whether it's possible to preserve redexes in user-written types and not to
+produce new ones while encoding the types. And the answer is "yes".
+
+But read Note [Spiney API] first.
+-}
+
+{- Note [Spiney API]
+Encoding of n-ary pattern functors into 1-ary pattern functors is hidden behind an API that pretends
+our types are in head-spine form. See @plutus/docs/fomega/deep-isorecursive/alternatives.md@ for
+details and discussion about the head-spine form approach.
+
+The reasons for providing such API are
+
+1. it's simple
+2. it hides all the gory details in such a way that we can change the representation of types and
+not change the API. For example, we can encode pattern functors in different ways (and we, in fact,
+do this) or we even can have the head-spine form in the AST and that wouldn't affect the API
+
+We could have an API like this: the user provides an n-ary pattern functor and we manipulate the AST
+directly which may or may not involve deconstruction of the AST depending on how we perform encoding.
+However the user might provide something that is not a pattern functor, but computes to a pattern
+functor and everything becomes more complicated. Instead, we require that the user provides the name
+of the data type being defined, a list of type variables (the ones that the pattern functor binds)
+along with their kinds and the body of the pattern functor separately. Having this information is
+enought to perform whatever encoding we want. Here is how it looks on the 'interlist' example:
+
+the n-ary pattern functor of 'interlist' is
+
+    \(interlist :: * -> * -> *) (a :: *) (b :: *) ->
+        all (r :: *). r -> (a -> b -> interlist b a -> r) -> r
+
+and we require the user (where "the user" means someone generating Plutus Core or writing it directly,
+i.e. either someone writing a compiler to Plutus Core or one of the creators of the language) to split
+this type into three components
+
+1. "interlist"         -- the name of the data type
+2. @[a :: *, b :: *]@  -- other type variables the pattern functors binds along with their kinds
+3. @all (r :: *). r -> (a -> b -> interlist b a -> r) -> r@  -- the body of the pattern functor
+
+and pass them to the 'makeRecursiveType' function (which also receives an annotation as its first
+argument just so that we have something to place in the AST when needed). Note that we do not require
+to provide the kind of 'interlist', because we can compute it from the kinds of other type variables.
+2
+The code constructing the data type itself:
```

Comment:
The example is there to connect ideas and the actual implementation via

```haskell
    makeRecursiveType () interlist [TyVarDecl () a $ Type (), TyVarDecl () b $ Type ()]
        . TyForall () r (Type ())  -- all (r :: *).
        . TyFun () nilElimTy       --     r ->
        . TyFun () consElimTy      --         (a -> b -> interlist b a -> r) ->
        $ TyVar () r               --             r
```

I think it's more helpful this way than just a reference. I could define the data type after the note, but in important modules I prefer to keep only the code that is actually needed.

Reply:
Yes, I wouldn't define it there, just point to the module where it is defined. I'm just wary of having substantial chunks of code in comments as they tend to go out of date even more quickly than comments themselves.

Reply:
> Yes, I wouldn't define it there, just point to the module where it is defined. I'm just wary of having substantial chunks of code in comments as they tend to go out of date even more quickly than comments themselves.

While I do agree with what you're saying, I believe it's better to sacrifice modularity here, because I think this example is helpful and referring to some other module makes the reference rather pointless, because jumping between modules is not something you want to do while reading docs.

How about the following: I'll just make a comment in the file that defines `interlist` in Plutus Core saying "if you change this definition, then also change the one in Note [Spiney API] in .../StdLib/Type.hs`. Not that I think this is a good practice, but I do want to have an example in the docs.